### ISSUE TITLE
Separating output plugins from streams (aka. graylog-server federation)

### ISSUE THREAD

**Issue by henrikjohansen** [Graylog2/graylog2-server#1004]
In order to build a federation pipeline ie. splitting the work over distinct clusters or groups of nodes within a cluster I would love to be able to forward all messages a given node receives without having to resort to streams.

Furthermore it would be awesome if you could actively choose whether or not to disable the Elasticsearch output completely or simply run with multiple, simultaneous outputs (it makes sense to have nodes that forward all events to riemann, graphite, etc while still indexing data into ES). 

For other use cases it makes sense to simply disable the ES output altogether because nodes further down the pipeline will take care of the indexing.

This would allow us to have dedicated nodes for receiving and extracting data, stream matching, indexing, etc that can be scaled individually and also allow us to build out cross datacenter operations like having nodes in each data center that process events and forward to central indexing nodes over TCP more intelligently.

**Comment by bernd** [comment]
:+1:

**Comment by dfch** [comment]
:+1: for the proposed option to disable ElasticSearch output completely (maybe the ES output could be implemented as an optional plugin).

@henrikjohansen why don't you create a stream with the condition `field_exists(message)` (which is always true) and attach an output to that stream? We do this to send all messages from a node to an upstream server (for audit trail purposes). Or am I missing something here?

**Comment by henrikjohansen** [comment]
I consider that to be somewhat of a hack - the architectural issue is that output are tied to streams and that's what I would like to see changed.

**Comment by kroepke** [comment]
FWIW the default output is already pluggable, even though it is not advertised as such.

It's entirely possible to replace the default Elasticsearch output with another output, but certain actions need to be performed internally, which we currently consider implementation details that might change without much notice, thus we haven't documented the procedure yet.
Additionally, other parts of the system expect messages to end up in Elasticsearch in order to work, so bypassing the ES output completely will result on various weird behavior.

If you really wanted to explore this further now, set the config option `default_message_output_class` to an implementation of `MessageOutput`.
Refer to the implementation of `https://github.com/Graylog2/graylog2-server/blob/1.0/graylog2-server/src/main/java/org/graylog2/outputs/ElasticSearchOutput.java` for the minimal set of methods and their behavior, the single most important action is calling `journal.markJournalOffsetCommitted(message.getJournalOffset());` on the messages that have been processed (with or without error), as this controls the commit marker in the journal. If you do not do this, your journal will grow up to the specified size limit and on restart would be replayed completely, every time.

However, please be advised that any of this behavior might change without notice in an upcoming version.

**Comment by kroepke** [comment]
Of course it would also be possible to create an output that forwards messages to more than one output (somewhat like `tee`). There are some timing issues to overcome with this, though, so please only use it in experimental setups.

**Comment by henrikjohansen** [comment]
Well - the question is if we're going to see the current model where all server nodes are equal (with the distinction of a 'master' for running certain jobs) in the future or if we'll be able to utilise different types / roles / classes of nodes such a processing node, storage nodes, etc.

If Graylog Inc is pursuing the latter this issues makes sense for a boarder audience - if not it should probably be closed to minimise the noise.

**Comment by kroepke** [comment]
I don't think we have made a clear decision on that yet, so I'd prefer to keep this open for now.
Proper federation needs a bit more than just forwarding messages, I guess, especially around failover and reliability. Just to a have issue number to refer to is good, imho.
